File: module-ocr/src/main/java/com/tokyonth/module_ocr/analyze/TextRecognitionAnalyzer.java
/*
 * Copyright (C) Jenly, MLKit Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tokyonth.module_ocr.analyze;

import android.graphics.Bitmap;

import androidx.annotation.NonNull;
import androidx.camera.core.ImageProxy;

import com.google.mlkit.vision.common.InputImage;
import com.google.mlkit.vision.text.Text;
import com.google.mlkit.vision.text.TextRecognition;
import com.google.mlkit.vision.text.TextRecognizer;
import com.google.mlkit.vision.text.TextRecognizerOptionsInterface;
import com.google.mlkit.vision.text.latin.TextRecognizerOptions;
import com.tokyonth.module_core.AnalyzeResult;
import com.tokyonth.module_core.analyze.Analyzer;
import com.tokyonth.module_core.util.BitmapUtils;
import com.tokyonth.module_core.util.LogUtils;

/**
 * @author <a href="mailto:jenly1314@gmail.com">Jenly</a>
 */
public class TextRecognitionAnalyzer implements Analyzer<Text> {

    private final TextRecognizer mTextRecognizer;

    public TextRecognitionAnalyzer() {
        this(null);
    }

    public TextRecognitionAnalyzer(TextRecognizerOptionsInterface options) {
        if (options != null) {
            mTextRecognizer = TextRecognition.getClient(options);
        } else {
            mTextRecognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS);
        }
    }

    @Override
    public void analyze(@NonNull ImageProxy imageProxy, @NonNull OnAnalyzeListener<AnalyzeResult<Text>> listener) {
        try {
            final Bitmap bitmap = BitmapUtils.getBitmap(imageProxy);
            assert bitmap != null;
            InputImage inputImage = InputImage.fromBitmap(bitmap, 0);
            mTextRecognizer.process(inputImage)
                    .addOnSuccessListener(result -> listener.onSuccess(new AnalyzeResult<>(bitmap, result)))
                    .addOnFailureListener(e -> listener.onFailure());
        } catch (Exception e) {
            LogUtils.w(e);
        }
    }

}

File: module-ocr/src/main/java/com/tokyonth/module_ocr/TextRecognitionAnalyzer.kt
package com.tokyonth.module_ocr

import android.content.ContentResolver
import com.google.mlkit.vision.text.TextRecognizer
import androidx.camera.core.ImageProxy
import com.tokyonth.module_core.analyze.Analyzer.OnAnalyzeListener
import com.tokyonth.module_core.AnalyzeResult
import android.graphics.Bitmap
import com.tokyonth.module_core.util.BitmapUtils
import android.net.Uri
import java.io.IOException
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.Text
import java.lang.Exception
import com.google.mlkit.vision.text.TextRecognition
import com.tokyonth.module_core.analyze.Analyzer
import com.tokyonth.module_ocr.factory.RecognizerFactory

class TextRecognitionAnalyzer constructor(
    private val contentResolver: ContentResolver,
    options: RecognizerFactory
) : Analyzer<Text?> {

    private var mTextRecognizer: TextRecognizer? = null

    init {
        mTextRecognizer = TextRecognition.getClient(options.createClient())
    }

    override fun analyze(
        imageProxy: ImageProxy,
        listener: OnAnalyzeListener<AnalyzeResult<Text?>>
    ) {
        analyzeBitmap(BitmapUtils.getBitmap(imageProxy), listener)
    }

    override fun analyze(
        uri: Uri, listener:
        OnAnalyzeListener<AnalyzeResult<Text?>>
    ) {
        try {
            analyzeBitmap(BitmapUtils.getBitmapFromContentUri(contentResolver, uri), listener)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun analyzeBitmap(bitmap: Bitmap?, listener: OnAnalyzeListener<AnalyzeResult<Text?>>) {
        try {
            if (bitmap == null)
                return
            val inputImage = InputImage.fromBitmap(bitmap, 0)
            mTextRecognizer!!.process(inputImage)
                .addOnSuccessListener { result: Text ->
                    listener.onSuccess(
                        AnalyzeResult(
                            bitmap,
                            result
                        )
                    )
                }
                .addOnFailureListener { listener.onFailure() }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

}
